### ISSUE TITLE
guide bot disabled in coop mode

### ISSUE THREAD

**Issue by cfeuersaenger** [dxx-rebirth/dxx-rebirth#364]
thanks for maintaining this cool old game! I found it yesterday and I am amazed.

However, I am missing the guide bot in coop mode. Since I am playing on linux anyway, I looked into the code and realized that it was disabled as part of the multiplayer code. It appears to be simple to reactivate it for coop mode and I am attaching a patch which appears to work for the first level and for a couple of commands (intensive tests will show if there are any issues left). Or was there some bigger issue with the guide bot and multiplayer?

[guidebotmultiplayer.txt](https://github.com/dxx-rebirth/dxx-rebirth/files/1671584/guidebotmultiplayer.txt)

**Comment by cfeuersaenger** [comment]
... however, old savegames can no longer be loaded with my suggested fix; it says 

similar/main/state.cpp:2002: BUG: object 0x891c08 has type 12, expected 4
similar/main/state.cpp:2002: BUG: object 0x895f18 has type 12, expected 4

and the client disconnects. 

The guide bot reappears for the host which appears to be the root cause of the problem.

**Comment by vLKp** [comment]
As I read the code, the guidebot has been absent from multiplayer since 1995-10-12 (or possibly 1995-12-10, depending on whether Matt Toschlog uses American (MM/DD/YY) or European (DD/MM/YY) date ordering - probably American, so -10-12).  In any case, the code to remove the guidebot was added with the comments:
```
main/gameseq.c:229://added 10/12/95: delete buddy bot if coop game.  Probably doesn't really belong here. -MT
main/gameseq.c:263:         obj_delete(i);      //kill the buddy in netgames
```

I am not aware of anyone who has ever tried to enable it, nor what issues it might cause.  Since it was never designed for multiplayer, I would not be surprised if the guidebot does not synchronize well across machines, especially if the player begins actively managing the guidebot's objectives.

---
The diagnostic you reported is a warning, not a fatal error.  On its own, it does not break anything.  I pushed a fix to silence that diagnostic, since it is harmless in this case.  Fields were loaded in the wrong order.  If your savegames are broken, something else is wrong.

You say the guidebot reappears for the host.  Does that mean it does not appear for the guest?  Were both systems running the patched build?

**Comment by cfeuersaenger** [comment]
Thanks for looking into this issue.

I took my patched executable and played three levels, and all worked well with the guide bot.

Once the client did not manage to reach the exit and failed to join afterwards for some reason (log file says "terminate called after throwing an instrance of 'multi::level_checksum_mismatch').

However, all others worked well. I even loaded the savegame that I claimed was broken -- and it was loaded correctly, client joined successfully, and both players finished the level. I remember that my last attempt actually contained the guide bot in the loaded mission... perhaps that was some intermediate state? I cannot remember.

The next level had the guide bot in its cage.

In other words: sounds good so far!

Some remarks:

* Note that my patch actually contains two further lines in escort.cpp. These are necessary in order to let the guide bot chat with you and in order to activate the guide bot controls (Shift F4).
* the guide bot only listens to the first player (i.e. the host). As soon as the first player exits the level, it appears to go to the second one (may need to be confirmed). But it is absolutely fine with me if the guide bot only talks to the host.
* I encountered no sync issues with the guide bot. The client sees it (including its messages).

**Comment by vLKp** [comment]
The guidebot extensively uses `ConsoleObject` to access player data.  If I recall correctly, this always points to the local player, not to the first player.  If so, then when certain player data, such as keys held, goes out of sync (which is natural, since players are not required to always grab all the same keys), different players will observe the guidebot making different decisions about its next goal.  Consider a case where the host always grabs all keys, and guests never grab any keys.  Guests would then observe the guidebot setting its goal to the first key; the host would observe the guidebot setting its goal to the reactor/boss.

I am not opposed to enabling the guidebot in multiplayer, but I think reliable operation will require more invasive changes than the patch you proposed.

**Comment by cfeuersaenger** [comment]
Our tests revealed no problems whatsoever -- except that the guide bot's movements are jumpy as it is lacks a proper integration into some smoothing/replication of the network code. Nonetheless, it functioned properly on both slave and host.

I also made a test in which the first player grabbed all keys are exited the level (all with the help of the guide bot). Afterwards, the guide bot tried to return to the second player which had no keys at all. I believe the guide bot was even silent in this case. That appears to be expected behavior to me.

Would it make sense to introduce some config option of sorts "[x]  guide bot in Coop [experimental]"? Maybe others are happy to use it as well. If it turns out to result in problems, it can be disabled.

**Comment by vLKp** [comment]
The first player does not need to exit the level.  Each player will perceive the guidebot as being loyal to him/her and picking its goals accordingly, regardless of what other players are on the level.  This would need to be fixed by changing the guidebot to pick its master from data that is synchronized among the peers.

Yes, if enabled, it would definitely be behind a configuration option.

**Comment by cfeuersaenger** [comment]
I see; you are referring to the guide-bot related stuff done in do_ai_frame() (ai.cpp).

I'll experiment with one of the following two attempts:
a) disable the decision making if the local player is not the master OR
b) change the logic to be based on player_num = 0. This requires some more attention as all arguments need to be computed for player_num = 0 if it is done by the client.

Solution (a) worked on first glance, but the network replication causes the guide bot to jump around on the client machine. At least it was smooth on the host machine.

Thanks for the pointers. I will reach out to you in case I find a suitable solution.

**Comment by vLKp** [comment]
I think option (b) would smooth out most of the problems.  If you want to be elaborate, make it track a new variable that initializes to player 0, but has some mechanism to switch it (on all machines) to refer to some other player.  If you do this, be careful to handle the case that the controlling player exits.  Player 0 can never quit without ending the game, but other players can leave and rejoin at any time.

If you have further questions, please ask.  I haven't wanted to spend the time to implement this (as yours is the only request I recall on this topic), but I'm happy to help you develop it.

**Comment by Diamond-Wolf** [comment]
My two cents:
What if the guidebot was tracked individually per player - each player sees its own instance as of the moment it was released. That way you could still get out of some crazy maze hunting for the blue key with GB's help so you stand a chance of exiting as the host gets the GB's help with the reactor.

**Comment by vLKp** [comment]
The guidebot can, and often does, open doors.  It would be very strange if other player's guidebots opened doors you could see, while your guidebot was far away.  The mine would appear to be haunted.  Yet it would be even stranger if doors opened by foreign guidebots did not open on the local screen, since the player who owned the guidebot could fly through it and others could not.

**Comment by cfeuersaenger** [comment]
I am not quite sure on how to understand the comments... do you advise me against some coarse of action?

I fully agree that tracking the guide bot per player, especially without visibility to other players, causes lots of unwanted side effects and would be, in effect, a show-stopper bug. 

I suppose it is much better to have a guide bot which listens to just one player (and, perhaps, switches over to the next one once the current player left the mine). Since it is a coop game, you may want to stick together anyway so you won't even realize it is not coming to you. 

In fact, my patch as it is attached to this ticket appears to do the job -- with the exception that the local player tries to send commands which are almost immediately overruled by the network replication code, such that the GB appears to jumps around (although it always does what the host says). Thus, it is just a matter of smoothing its movements.

**Comment by vLKp** [comment]
My most recent comment was in response to Diamond-Wolf suggesting that the game allow the guidebot to desynchronize.

Your patch is insufficient because the guidebot, as coded, will always desynchronize if the players do not keep their ships in similar state.  When the guidebot checks whether to return to the player, it always uses the local player, not a mutually agreed master.  When the guidebot guesses a goal based on collected keys, it inspects the local player, not a mutually agreed master.  Every time the guidebot needs to make any decision that involves any player state, it always uses the local player.  The network code can smooth over some of this by forcibly resynchronizing its position, but a full implementation needs the guidebot to settle on a single player as its master.  Fortunately, it consistently uses `ConsoleObject` to get the master.  Fixing this is a matter of identifying guidebot-specific uses of `ConsoleObject` and converting them to something that will remain synchronized in a network game.

The existing desynchronization susceptible version might be an acceptable limitation for a small group playing with a local patch.  I do not want to publish such a change generally, because I expect that players will expect the feature to work without obvious desynchronization issues.

**Comment by vLKp** [comment]
I recently revisited this request, and wrote up an implementation that I think will work better in multiplayer.  Please test and report any problems.
